File: android/core/network/src/main/kotlin/xyz/klinik/network/ApiError.kt
package xyz.klinik.network

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonPrimitive

/** The error body every backend failure uses. */
@Serializable
data class ErrorResponse(
    @SerialName("statusCode") val statusCode: Int = 0,
    @SerialName("message") private val rawMessage: JsonElement? = null,
    @SerialName("error") val error: String? = null,
) {
    /**
     * Validation failures arrive as an array of messages and everything else as
     * a single string; both reach the UI as one field.
     */
    val message: String
        get() = when (val element = rawMessage) {
            null -> ""
            is JsonPrimitive -> element.content
            else -> runCatching {
                element.jsonArray.joinToString("\n") { it.jsonPrimitive.content }
            }.getOrDefault("")
        }
}

/** Machine-readable codes the backend returns on the authentication path. */
enum class AuthErrorCode(val code: String) {
    INVALID_CREDENTIALS("INVALID_CREDENTIALS"),
    ACCOUNT_LOCKED("ACCOUNT_LOCKED"),
    ACCOUNT_INACTIVE("ACCOUNT_INACTIVE"),
    MFA_REQUIRED("MFA_REQUIRED"),
    MFA_INVALID("MFA_INVALID"),
    MFA_SETUP_REQUIRED("MFA_SETUP_REQUIRED"),
    INVITATION_INVALID("INVITATION_INVALID"),
    INVITATION_EXPIRED("INVITATION_EXPIRED"),
    INVITATION_ATTEMPTS_EXCEEDED("INVITATION_ATTEMPTS_EXCEEDED"),
    PASSWORD_TOO_WEAK("PASSWORD_TOO_WEAK"),
    ;

    companion object {
        fun from(value: String): AuthErrorCode? = entries.firstOrNull { it.code == value }
    }
}

sealed class ApiError(message: String? = null) : Exception(message) {
    /**
     * No usable connection. Distinct from a server failure because the UI
     * offers a different remedy (spec M15: the offline indicator).
     */
    data object Offline : ApiError("offline")
    data object TimedOut : ApiError("timed out")

    /** A recognised authentication outcome; the UI branches on the code. */
    data class Auth(val code: AuthErrorCode, val body: ErrorResponse) : ApiError(code.code)

    data class Unauthorized(val body: ErrorResponse) : ApiError("unauthorized")
    data class Forbidden(val body: ErrorResponse) : ApiError("forbidden")

    /**
     * Also returned for a record outside the caller's scope — the backend makes
     * the two indistinguishable on purpose, so the client must not tell the
     * user the record exists.
     */
    data class NotFound(val body: ErrorResponse) : ApiError("not found")

    data class Validation(val body: ErrorResponse) : ApiError("validation")
    data class Conflict(val body: ErrorResponse) : ApiError("conflict")
    data class RateLimited(val retryAfterSeconds: Long?) : ApiError("rate limited")
    data class Server(val body: ErrorResponse) : ApiError("server")
    data class Decoding(val detail: String) : ApiError("decoding")
    data class Unknown(val status: Int) : ApiError("unknown")

    /** Whether retrying the same request unchanged could plausibly succeed. */
    val isRetryable: Boolean
        get() = this is Offline || this is TimedOut || this is RateLimited || this is Server

    /** Whether the session is over and the user has to sign in again. */
    val requiresReauthentication: Boolean
        get() = when (this) {
            is Unauthorized -> true
            is Auth -> code == AuthErrorCode.ACCOUNT_INACTIVE || code == AuthErrorCode.ACCOUNT_LOCKED
            else -> false
        }

    companion object {
        /** Maps a response to the case the UI branches on. */
        fun from(status: Int, body: ErrorResponse, retryAfterSeconds: Long? = null): ApiError {
            AuthErrorCode.from(body.message)?.let { return Auth(it, body) }

            return when (status) {
                400, 422 -> Validation(body)
                401 -> Unauthorized(body)
                403 -> Forbidden(body)
                404 -> NotFound(body)
                409 -> Conflict(body)
                429 -> RateLimited(retryAfterSeconds)
                in 500..599 -> Server(body)
                else -> Unknown(status)
            }
        }
    }
}

File: ios/Sources/KlinikCore/APIError.swift
import Foundation

/// The error body every backend failure uses.
public struct ErrorResponse: Codable, Sendable, Equatable {
    public let statusCode: Int
    public let message: String
    public let error: String?

    /// How long a lockout has left. Present only on ACCOUNT_LOCKED.
    public let retryAfterSeconds: Int?

    private enum CodingKeys: String, CodingKey {
        case statusCode, message, error, retryAfterSeconds
    }

    public init(
        statusCode: Int,
        message: String,
        error: String? = nil,
        retryAfterSeconds: Int? = nil
    ) {
        self.statusCode = statusCode
        self.message = message
        self.error = error
        self.retryAfterSeconds = retryAfterSeconds
    }

    public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        statusCode = try container.decodeIfPresent(Int.self, forKey: .statusCode) ?? 0
        error = try container.decodeIfPresent(String.self, forKey: .error)
        retryAfterSeconds = try container.decodeIfPresent(Int.self, forKey: .retryAfterSeconds)

        // Validation failures arrive as an array of messages; everything else
        // as a single string. Both reach the UI as one field.
        if let single = try? container.decode(String.self, forKey: .message) {
            message = single
        } else if let many = try? container.decode([String].self, forKey: .message) {
            message = many.joined(separator: "\n")
        } else {
            message = ""
        }
    }
}

/// Machine-readable codes the backend returns on the authentication path.
public enum AuthErrorCode: String, Sendable {
    case invalidCredentials = "INVALID_CREDENTIALS"
    case accountLocked = "ACCOUNT_LOCKED"
    case accountInactive = "ACCOUNT_INACTIVE"
    case mfaRequired = "MFA_REQUIRED"
    case mfaInvalid = "MFA_INVALID"
    case mfaSetupRequired = "MFA_SETUP_REQUIRED"
    case invitationInvalid = "INVITATION_INVALID"
    case invitationExpired = "INVITATION_EXPIRED"
    case invitationAttemptsExceeded = "INVITATION_ATTEMPTS_EXCEEDED"
    case passwordTooWeak = "PASSWORD_TOO_WEAK"
}

public enum APIError: Error, Sendable, Equatable {
    /// No usable connection. Distinct from a server failure because the UI
    /// offers a different remedy (spec M15: the offline indicator).
    case offline
    case timedOut
    /// Recognised authentication outcome; the UI branches on the code.
    case auth(AuthErrorCode, ErrorResponse)
    case unauthorized(ErrorResponse)
    case forbidden(ErrorResponse)
    /// Also returned for a record outside the caller's scope — the backend
    /// deliberately makes the two indistinguishable, so the client must not
    /// tell the user the record exists.
    case notFound(ErrorResponse)
    case validation(ErrorResponse)
    case conflict(ErrorResponse)
    case rateLimited(retryAfter: TimeInterval?)
    case server(ErrorResponse)
    case decoding(String)
    case unknown(status: Int)

    /**
     * The write could not be delivered and was kept to be sent later.
     *
     * Thrown rather than returned because there is no answer to return: the
     * server has not seen the change, so there is no record, no id and no
     * timestamp to hand back. The screen that catches this tells the user
     * their work is saved and not yet sent — which is the truth, and is
     * neither the success nor the failure the other cases describe.
     */
    case queuedForLater

    /**
     * Whether the clinic was unreachable, as opposed to unwilling.
     *
     * The distinction decides whether a stale copy may be shown: a server that
     * answered "no" gave an answer, and replacing it with yesterday's "yes"
     * would be the client overruling the clinic.
     */
    public var isConnectivity: Bool {
        self == .offline || self == .timedOut
    }

    /// Whether retrying the same request unchanged could plausibly succeed.
    public var isRetryable: Bool {
        switch self {
        case .offline, .timedOut, .rateLimited, .server:
            return true
        default:
            return false
        }
    }

    /// Whether the session is over and the user has to sign in again.
    public var requiresReauthentication: Bool {
        switch self {
        case .unauthorized:
            return true
        case .auth(let code, _):
            return code == .accountInactive || code == .accountLocked
        default:
            return false
        }
    }
}

public extension APIError {
    /// Maps a response to the case the UI branches on.
    static func from(status: Int, body: ErrorResponse, retryAfter: TimeInterval? = nil) -> APIError {
        if let code = AuthErrorCode(rawValue: body.message) {
            return .auth(code, body)
        }

        switch status {
        case 400, 422: return .validation(body)
        case 401: return .unauthorized(body)
        case 403: return .forbidden(body)
        case 404: return .notFound(body)
        case 409: return .conflict(body)
        case 429: return .rateLimited(retryAfter: retryAfter)
        case 500...599: return .server(body)
        default: return .unknown(status: status)
        }
    }
}

public extension APIError {
    /// Whether the server refused, as opposed to failing to answer at all.
    var isRefusal: Bool {
        if case .forbidden = self { return true }

        return false
    }
}

/// A best-effort read: what came back, or why nothing did.
///
/// `try?` throws the reason away, and several screens then read "nothing came
/// back from any of these" as "this account may not see the panel". That is
/// true of a 403 and of nothing else — an unreachable clinic, a 500, or a
/// response the app cannot parse would all be reported to a finance officer as
/// a permission problem, on a screen with no way to try again.
public enum Attempted<Value: Sendable>: Sendable {
    case arrived(Value)
    case failed(APIError)

    public var value: Value? {
        if case .arrived(let value) = self { return value }

        return nil
    }

    public var error: APIError? {
        if case .failed(let error) = self { return error }

        return nil
    }
}

/// Runs a read that is allowed to fail, keeping the reason.
public func attempt<Value: Sendable>(
    _ work: @Sendable () async throws -> Value
) async -> Attempted<Value> {
    do {
        return .arrived(try await work())
    } catch let error as APIError {
        return .failed(error)
    } catch {
        return .failed(.unknown(status: 0))
    }
}

/// What a screen should show when every one of its reads came back empty.
public enum ReadOutcome: Sendable, Equatable {
    /// Every failure was a refusal: this account may not see the screen.
    case refused
    /// Something else went wrong. Worth saying what, and worth a retry.
    case failed(String)

    public static func of(_ errors: [APIError]) -> ReadOutcome {
        guard let reason = errors.first(where: { !$0.isRefusal }) else { return .refused }

        return .failed(L10n.message(for: reason))
    }
}
